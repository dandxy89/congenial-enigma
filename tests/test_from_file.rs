use std::path::PathBuf;

use lp_parser_rs::{
    model::{lp_error::LPParserError, lp_problem::LPProblem},
    parse::{parse_file, parse_lp_file},
};

#[macro_export]
macro_rules! generate_test {
    ($test_name:ident, $file:expr, $o_sum:expr, $c_sum:expr) => {
        #[test]
        fn $test_name() {
            let result = read_file_from_resources($file).unwrap();

            #[cfg(feature = "serde")]
            insta::assert_yaml_snapshot!(result, {
                ".variables" => insta::sorted_redaction(),
                ".objectives" => insta::sorted_redaction(),
                ".constraints" => insta::sorted_redaction(),
            });

            let summation = &result.objectives.iter().map(|o| o.coefficients.iter().map(|c| c.coefficient).sum::<f64>()).sum();
            float_eq::assert_float_eq!($o_sum, *summation, abs <= 1e-2);

            let summation = &result.constraints.iter().map(|(_, c)| c.coefficients().iter().map(|c| c.coefficient).sum::<f64>()).sum();
            float_eq::assert_float_eq!($c_sum, *summation, abs <= 1e-2);
        }
    };
}

generate_test!(afiro, "afiro.lp", -4.100, 25.369);
generate_test!(afiro_ext, "afiro_ext.lp", -5.239, 25.369);
generate_test!(boeing1, "boeing1.lp", 1187.985, 194612.346);
generate_test!(boeing2, "boeing2.lp", 78.488, 20863.836);
generate_test!(fit1d, "fit1d.lp", 82457.0, -146871.180);
generate_test!(kb2, "kb2.lp", 11.675, 10143.724);
generate_test!(pulp, "pulp.lp", 1., 73.44);
generate_test!(pulp2, "pulp2.lp", -6.0, 147.);
generate_test!(sc50a, "sc50a.lp", -1., 30.3);
generate_test!(model2, "model2.lp", 0., 6.);
generate_test!(limbo, "limbo.lp", 2., 0.);
generate_test!(obj3_2cons, "3obj_2cons.lp", 1., -186.);
generate_test!(obj_2cons_only_binary_vars, "2obj_2cons_only_binary_vars.lp", -7.5, -186.);
generate_test!(obj_2cons_all_variable_types, "2obj_2cons_all_variable_types.lp", -7.5, -186.);
generate_test!(obj_1cons_all_variables_with_bounds, "1obj_1cons_all_variables_with_bounds.lp", -1., 16.5);
generate_test!(semi_continuous, "semi_continuous.lp", 2., 0.);
generate_test!(sos, "sos.lp", 0., 17.5);
generate_test!(test, "test.lp", 2., 2.9899);
generate_test!(test2, "test2.lp", -6., 147.);
generate_test!(empty_bounds, "empty_bounds.lp", 11., 2.);
generate_test!(blank_lines, "blank_lines.lp", 11., 2.);
generate_test!(optional_labels, "optional_labels.lp", 11., 1.);
generate_test!(infile_comments, "infile_comments.lp", 43., 7.);
generate_test!(infile_comments2, "infile_comments2.lp", 43., 0.);
generate_test!(missing_signs, "missing_signs.lp", 43., -3.);
generate_test!(fit2d, "fit2d.lp", 349048.9, -296677.389);
generate_test!(scientific_notation, "scientific_notation.lp", 238.0324, -43.87680);

#[test]
fn invalid() {
    let result = read_file_from_resources("invalid.lp");
    assert!(result.is_err());
}

fn read_file_from_resources(file_name: &str) -> Result<LPProblem, LPParserError> {
    let mut file_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    file_path.push(format!("resources/{file_name}"));
    let contents = parse_file(&file_path)?;
    parse_lp_file(&contents)
}
