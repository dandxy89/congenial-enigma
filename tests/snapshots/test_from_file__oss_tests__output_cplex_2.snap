---
source: tests/test_from_file.rs
expression: result
---
problem_name: ~
problem_sense: Minimize
variables:
  X0: Integer
  X1: Integer
  X2: Integer
  X3: Integer
  X4: Integer
  X5: Integer
  X6: Integer
  X7: Integer
  X8: Integer
  X9: Integer
objectives:
  - name: obj
    coefficients:
      - var_name: X9
        coefficient: 1
constraints:
  con_2024:
    Standard:
      name: con_2024
      coefficients:
        - var_name: X0
          coefficient: -1
      sense: ">="
      rhs: -5
  con_2025:
    Standard:
      name: con_2025
      coefficients:
        - var_name: X0
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2026:
    Standard:
      name: con_2026
      coefficients:
        - var_name: X0
          coefficient: -1
        - var_name: X3
          coefficient: -1
        - var_name: X6
          coefficient: -1
      sense: ">="
      rhs: -7
  con_2027:
    Standard:
      name: con_2027
      coefficients:
        - var_name: X0
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2028:
    Standard:
      name: con_2028
      coefficients:
        - var_name: X1
          coefficient: -1
      sense: ">="
      rhs: -3
  con_2029:
    Standard:
      name: con_2029
      coefficients:
        - var_name: X1
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2030:
    Standard:
      name: con_2030
      coefficients:
        - var_name: X1
          coefficient: -1
        - var_name: X4
          coefficient: -1
        - var_name: X7
          coefficient: -1
      sense: ">="
      rhs: -3
  con_2031:
    Standard:
      name: con_2031
      coefficients:
        - var_name: X1
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2032:
    Standard:
      name: con_2032
      coefficients:
        - var_name: X1
          coefficient: -1
        - var_name: X4
          coefficient: -2
        - var_name: X0
          coefficient: -3
        - var_name: X7
          coefficient: -3
        - var_name: X8
          coefficient: -3
        - var_name: X3
          coefficient: -4
        - var_name: X5
          coefficient: -4
        - var_name: X2
          coefficient: -9999
        - var_name: X6
          coefficient: -9999
        - var_name: X9
          coefficient: 1
      sense: "="
      rhs: 0
  con_2033:
    Standard:
      name: con_2033
      coefficients:
        - var_name: X2
          coefficient: -1
      sense: ">="
      rhs: -5
  con_2034:
    Standard:
      name: con_2034
      coefficients:
        - var_name: X2
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2035:
    Standard:
      name: con_2035
      coefficients:
        - var_name: X2
          coefficient: 1
        - var_name: X1
          coefficient: 1
        - var_name: X0
          coefficient: 1
      sense: "="
      rhs: 5
  con_2036:
    Standard:
      name: con_2036
      coefficients:
        - var_name: X2
          coefficient: -1
        - var_name: X5
          coefficient: -1
        - var_name: X8
          coefficient: -1
      sense: ">="
      rhs: -5
  con_2037:
    Standard:
      name: con_2037
      coefficients:
        - var_name: X2
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2038:
    Standard:
      name: con_2038
      coefficients:
        - var_name: X3
          coefficient: -1
      sense: ">="
      rhs: -7
  con_2039:
    Standard:
      name: con_2039
      coefficients:
        - var_name: X3
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2040:
    Standard:
      name: con_2040
      coefficients:
        - var_name: X3
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2041:
    Standard:
      name: con_2041
      coefficients:
        - var_name: X4
          coefficient: -1
      sense: ">="
      rhs: -3
  con_2042:
    Standard:
      name: con_2042
      coefficients:
        - var_name: X4
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2043:
    Standard:
      name: con_2043
      coefficients:
        - var_name: X4
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2044:
    Standard:
      name: con_2044
      coefficients:
        - var_name: X5
          coefficient: -1
      sense: ">="
      rhs: -5
  con_2045:
    Standard:
      name: con_2045
      coefficients:
        - var_name: X5
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2046:
    Standard:
      name: con_2046
      coefficients:
        - var_name: X5
          coefficient: 1
        - var_name: X4
          coefficient: 1
        - var_name: X3
          coefficient: 1
      sense: "="
      rhs: 7
  con_2047:
    Standard:
      name: con_2047
      coefficients:
        - var_name: X5
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2048:
    Standard:
      name: con_2048
      coefficients:
        - var_name: X6
          coefficient: -1
      sense: ">="
      rhs: -3
  con_2049:
    Standard:
      name: con_2049
      coefficients:
        - var_name: X6
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2050:
    Standard:
      name: con_2050
      coefficients:
        - var_name: X6
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2051:
    Standard:
      name: con_2051
      coefficients:
        - var_name: X7
          coefficient: -1
      sense: ">="
      rhs: -3
  con_2052:
    Standard:
      name: con_2052
      coefficients:
        - var_name: X7
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2053:
    Standard:
      name: con_2053
      coefficients:
        - var_name: X7
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2054:
    Standard:
      name: con_2054
      coefficients:
        - var_name: X8
          coefficient: -1
      sense: ">="
      rhs: -3
  con_2055:
    Standard:
      name: con_2055
      coefficients:
        - var_name: X8
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2056:
    Standard:
      name: con_2056
      coefficients:
        - var_name: X8
          coefficient: 1
        - var_name: X7
          coefficient: 1
        - var_name: X6
          coefficient: 1
      sense: "="
      rhs: 3
  con_2057:
    Standard:
      name: con_2057
      coefficients:
        - var_name: X8
          coefficient: 1
      sense: ">="
      rhs: 0
  con_2058:
    Standard:
      name: con_2058
      coefficients:
        - var_name: X9
          coefficient: -1
      sense: ">="
      rhs: -80082
  con_2059:
    Standard:
      name: con_2059
      coefficients:
        - var_name: X9
          coefficient: 1
      sense: ">="
      rhs: 0
